File: next/internal/statistics/statistics.go
package statistics

import (
	"math"
	"sort"
	"time"

	"netwatcher/next/internal/domain"
)

func Build(hours int, measurements []domain.Measurement, outages []domain.Outage) domain.Statistics {
	if hours <= 0 {
		hours = 24
	}
	from := time.Now().Add(-time.Duration(hours) * time.Hour)
	to := time.Now()
	type bucket struct {
		name, host, mode string
		values           []float64
		success, total   int
	}
	buckets := map[string]*bucket{}
	all := make([]float64, 0)
	successful := 0
	for _, m := range measurements {
		b := buckets[m.TargetID]
		if b == nil {
			b = &bucket{name: m.TargetName, host: m.Host, mode: m.Mode}
			buckets[m.TargetID] = b
		}
		b.total++
		if m.Success {
			b.success++
			successful++
			b.values = append(b.values, m.Latency)
			all = append(all, m.Latency)
		}
	}
	result := domain.Statistics{RangeHours: hours, From: from.Format(time.RFC3339), To: to.Format(time.RFC3339), Samples: len(measurements), Successful: successful}
	if result.Samples > 0 {
		result.PacketLoss = 100 * float64(result.Samples-successful) / float64(result.Samples)
		result.Uptime = 100 * float64(successful) / float64(result.Samples)
	}
	result.AverageLatency = avg(all)
	result.P95Latency = percentile(all, .95)
	targetJitters := make([]float64, 0, len(buckets))
	for _, bucket := range buckets {
		if len(bucket.values) >= 2 {
			targetJitters = append(targetJitters, jitter(bucket.values))
		}
	}
	result.Jitter = avg(targetJitters)
	for _, o := range outages {
		result.OutageCount++
		result.OutageSeconds += o.DurationSeconds
	}
	keys := make([]string, 0, len(buckets))
	for k := range buckets {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, id := range keys {
		b := buckets[id]
		s := domain.TargetStatistics{TargetID: id, TargetName: b.name, Host: b.host, Mode: b.mode, Samples: b.total, Successful: b.success}
		if b.total > 0 {
			s.PacketLoss = 100 * float64(b.total-b.success) / float64(b.total)
			s.Uptime = 100 * float64(b.success) / float64(b.total)
		}
		s.AverageLatency = avg(b.values)
		s.MinimumLatency = min(b.values)
		s.MaximumLatency = max(b.values)
		s.P95Latency = percentile(b.values, .95)
		s.Jitter = jitter(b.values)
		result.TargetBreakdown = append(result.TargetBreakdown, s)
	}
	return result
}
func avg(v []float64) float64 {
	if len(v) == 0 {
		return 0
	}
	t := 0.
	for _, x := range v {
		t += x
	}
	return t / float64(len(v))
}
func min(v []float64) float64 {
	if len(v) == 0 {
		return 0
	}
	m := v[0]
	for _, x := range v[1:] {
		if x < m {
			m = x
		}
	}
	return m
}
func max(v []float64) float64 {
	if len(v) == 0 {
		return 0
	}
	m := v[0]
	for _, x := range v[1:] {
		if x > m {
			m = x
		}
	}
	return m
}
func percentile(v []float64, p float64) float64 {
	if len(v) == 0 {
		return 0
	}
	c := append([]float64(nil), v...)
	sort.Float64s(c)
	i := int(math.Ceil(p*float64(len(c)))) - 1
	if i < 0 {
		i = 0
	}
	if i >= len(c) {
		i = len(c) - 1
	}
	return c[i]
}
func jitter(v []float64) float64 {
	if len(v) < 2 {
		return 0
	}
	d := make([]float64, 0, len(v)-1)
	for i := 1; i < len(v); i++ {
		d = append(d, math.Abs(v[i]-v[i-1]))
	}
	return avg(d)
}

File: rust_core/src/statistics.rs
use std::collections::BTreeMap;

use chrono::{DateTime, Utc};

use crate::models::{Measurement, Outage, Statistics, TargetStatistics};

#[derive(Default)]
struct TargetAccumulator {
    name: String,
    host: String,
    mode: String,
    samples: usize,
    successful: usize,
    latencies: Vec<f64>,
}

fn percentile95(values: &[f64]) -> f64 {
    if values.is_empty() {
        return 0.0;
    }
    let mut sorted = values.to_vec();
    sorted.sort_by(f64::total_cmp);
    let index = (((sorted.len() - 1) as f64) * 0.95).round() as usize;
    sorted[index.min(sorted.len() - 1)]
}

fn average(values: &[f64]) -> f64 {
    if values.is_empty() {
        0.0
    } else {
        values.iter().sum::<f64>() / values.len() as f64
    }
}

fn jitter(values: &[f64]) -> f64 {
    if values.len() < 2 {
        return 0.0;
    }
    values
        .windows(2)
        .map(|pair| (pair[1] - pair[0]).abs())
        .sum::<f64>()
        / (values.len() - 1) as f64
}

pub fn build(
    range_hours: i64,
    since: DateTime<Utc>,
    now: DateTime<Utc>,
    measurements: &[Measurement],
    outages: &[Outage],
) -> Statistics {
    let mut grouped: BTreeMap<String, TargetAccumulator> = BTreeMap::new();
    for measurement in measurements {
        let item = grouped.entry(measurement.target_id.clone()).or_default();
        item.name = measurement.target_name.clone();
        item.host = measurement.host.clone();
        item.mode = measurement.mode.clone();
        item.samples += 1;
        if measurement.success {
            item.successful += 1;
            item.latencies.push(measurement.latency);
        }
    }

    let mut breakdown = Vec::with_capacity(grouped.len());
    for (target_id, item) in grouped {
        let failures = item.samples.saturating_sub(item.successful);
        let packet_loss = if item.samples == 0 {
            0.0
        } else {
            failures as f64 / item.samples as f64 * 100.0
        };
        let minimum_latency = item
            .latencies
            .iter()
            .copied()
            .min_by(f64::total_cmp)
            .unwrap_or(0.0);
        let maximum_latency = item
            .latencies
            .iter()
            .copied()
            .max_by(f64::total_cmp)
            .unwrap_or(0.0);
        breakdown.push(TargetStatistics {
            target_id,
            target_name: item.name,
            host: item.host,
            mode: item.mode,
            samples: item.samples,
            successful: item.successful,
            packet_loss,
            average_latency: average(&item.latencies),
            minimum_latency,
            maximum_latency,
            p95_latency: percentile95(&item.latencies),
            jitter: jitter(&item.latencies),
            uptime: if item.samples == 0 {
                0.0
            } else {
                item.successful as f64 / item.samples as f64 * 100.0
            },
        });
    }

    let samples = measurements.len();
    let successful = measurements.iter().filter(|item| item.success).count();
    let latencies: Vec<f64> = measurements
        .iter()
        .filter(|item| item.success)
        .map(|item| item.latency)
        .collect();
    Statistics {
        range_hours,
        from: since.to_rfc3339(),
        to: now.to_rfc3339(),
        samples,
        successful,
        packet_loss: if samples == 0 {
            0.0
        } else {
            (samples - successful) as f64 / samples as f64 * 100.0
        },
        average_latency: average(&latencies),
        p95_latency: percentile95(&latencies),
        jitter: jitter(&latencies),
        uptime: if samples == 0 {
            0.0
        } else {
            successful as f64 / samples as f64 * 100.0
        },
        outage_count: outages.len(),
        outage_seconds: outages.iter().map(|item| item.duration_seconds).sum(),
        target_breakdown: breakdown,
    }
}
